packages/design-system/src/typings/svg.d.ts

```diff
@@ -0,0 +1,4 @@
+declare module '*.svg' {
+  const content: React.FunctionComponent<React.SVGAttributes<SVGElement>>;
+  export default content;
+}
```

Comment:
@swissspidy, this works partially. It works fine in my editor, but when running the TS workflow, I get these errors:

```
packages/design-system/dist-types/icons/index.d.ts:4:40 - error TS2307: Cannot find module './align_bottom.svg' or its corresponding type declarations.

4 export { default as AlignBottom } from './align_bottom.svg';
                                         ~~~~~~~~~~~~~~~~~~~~

packages/design-system/dist-types/icons/index.d.ts:5:40 - error TS2307: Cannot find module './align_center.svg' or its corresponding type declarations.

5 export { default as AlignCenter } from './align_center.svg';
                                         ~~~~~~~~~~~~~~~~~~~~
```

Etc for all 140 SVG's in the design system. Any idea how to make it work correctly?

Reply:
Does it work with a config like https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/story-editor/src/typings/svg-imports.d.ts?

Reply:
Still only in editor, workflow reports the some problem. Just pushed with this changed typing.

And it's actually not just the workflow that fails. The actual imports fail as well when running the editor. Webpack seems to be missing a loader or can't find the right one:

```
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> <svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
| <path d="M8.5 21H23.5" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round"/>
| <path fill-rule="evenodd" clip-rule="evenodd" d="M15.5 10.5C14.6716 10.5 14 11.1716 14 12V17C14 17.8284 14.6716 18.5 15.5 18.5H16.5C17.3284 18.5 18 17.8284 18 17V12C18 11.1716 17.3284 10.5 16.5 10.5H15.5Z" fill="currentColor"/>
```

Reply:
Ah, the loader issue was fixed by #12907, so should work when merging latest changes from main

Reply:
Hmm, it compiles with:

```ts
declare module '*.svg' {
  const content: string;
  export default content;
}
```

But then other things break, because they aren't actually string and we need to treat them like React elements (e.g. by adding class names and other properties to them elsewhere in the codebase).

If I do this:

```ts
/**
 * External dependencies
 */
import type { ReactElement, SVGProps } from 'react';

declare module '*.svg' {
  const Icon: (props: SVGProps<SVGElement>) => ReactElement;
  export default Icon;
}
```

It just fails without any errors or warnings (just that no corresponding type can be found). Not really sure why?

I also tried to swap in `SVGAttributes` for `SVGProps` (as it seems more correct), but no luck either.

Reply:
Hmm then I am not sure either 😕

Reply:
@swissspidy, any idea how to resolve this? My only other option, ChatGPT, doesn't know either 😭

Reply:
I fixed it! Well, [this answer](https://stackoverflow.com/questions/39040108/import-class-in-definition-file-d-ts) fixed my knowledge:

> Basically, TypeScript has two kind of module types declaration: "local" (normal modules) and ambient (global). The second kind allows to write global modules declaration that are merged with existing modules declaration. What are the differences between this files?
> `d.ts` files are treated as an ambient module declarations only if they don't have any imports. If you provide an import line, it's now treated as a normal module file, not the global one, so augmenting modules definitions doesn't work.

So this works (but stinks):

```ts
declare module '*.svg' {
  const Icon: (
    props: import('react').SVGAttributes<SVGElement>
  ) => import('react').ReactElement;
  export default Icon;
}
```

Alternatively we re-create this declaration in every folder that has SVG's, but there's actually quite a few, as several components have their own icons and graphics, and I'm still not convinced it would work.

So let's just keep this?

Reply:
Ah, it's actually as simple as just:

```ts
declare module '*.svg' {
  const Icon: (props: React.SVGAttributes<SVGElement>) => React.ReactElement;
  export default Icon;
}
```

The problem now is, that it doesn't work fully.

Yes, I can build the design system package correctly by running `npx tsc --build` inside the package. It runs correctly (with this file in place) and not without it.

But I cannot build the whole system, because `*.d.ts` files aren't copied to the `dist-types` output (they're only considered necessary **for** the build, not **part of** the build), so once the types have been summarized in dist-types, they're no longer "complete".

If I run `npm run workflow:bundle-packages:types` I get errors like:

```
packages/design-system/dist-types/images/index.d.ts:7:37 - error TS2307: Cannot find module './desert_bw.svg' or its corresponding type declarations.

  export { default as DesertBw } from "./desert_bw.svg";
```

And that makes sense, because once moved to the dist-types folder, then yes, the import no longer has a type for the SVG file. But how do we fix that? I have no clue...

Please halp, @swissspidy, you're my only hope! 🙏

Reply:
Unfortunately I'm at loss right now as well 😞 

I tried including the definition via `typeRoots` rather than `include`, and the only difference is that the error message is now for the source file and not for the one in `dist-types`. Progress I guess? But to me that sounds like it's not really an issue with copying over `d.ts` file.

The file I used:

```json5
{
  "extends": "../../tsconfig.shared.json",
  "compilerOptions": {
    "rootDir": "src",
    "declarationDir": "dist-types",
    "typeRoots": ["./src/typings", "../../node_modules/@types"]
  },
  "references": [
    { "path": "../i18n" },
    { "path": "../patterns" },
    { "path": "../react" },
    { "path": "../tracking" }
  ],
  "include": [
    "src/components",
    "src/contexts",
    "src/icons",
    "src/images",
    "src/theme",
    "src/types",
    "src/utils"
  ]
}
```

Then I also stumbled upon https://duncanleung.com/typescript-module-declearation-svg-img-assets/ and subsequently tried to move the type definition to a subfolder of `typings`, but to no avail.

I think I need to look at this again later with fresh eyes.

----

Aside: there are a couple of type errors related to `useDebouncedCallback` in `useDropDown` and `useSearch`

Reply:
I had an idea about creating a global `typings/` folder at the root and `/// <reference />` that, but it didn't seem to work. Almost, but then not again.

Another idea though: We create a new `svg` package with all the SVG files, and all other packages, that need SVG's import that package. Then we can use whatever hacks we need to make that package work, including perhaps manually crafting a `*.d.ts` if need be. Does that sound stupid?

Reply:
Can't believe that 6148ff51d589c96d7df5252b5bef189901b6cad8 solves the issue 😭 

The incorrect `declare module` definitions in `story-editor` were breaking this 🤷

Reply:
Oh wow, good find! But I though the inline-icons were actually imported as strings as used in various packages? But hey, if it works!

Reply:
Inline-icons is only used in the moveable package IIRC, so we could move that special case there